src/web/middleware/precondition.rs

```diff
@@ -76,140 +75,116 @@ where
+                    return Ok(sreq.into_response(
                         HttpResponse::Unauthorized()
                             .content_type("application/json")
                             .body("Invalid Authorization".to_owned())
                             .into_body(),
-                    ),
-                ));
-            }
-        };
-        let edb = extrude_db(&sreq.extensions());
-        let db = match edb {
-            Ok(v) => v,
-            Err(e) => {
-                error!("⚠️ Database access error {:?}", e);
-                queue_report(sreq.extensions_mut(), &e);
-                return Box::pin(future::ok(
-                    sreq.into_response(
-                        HttpResponse::InternalServerError()
-                            .content_type("application/json")
-                            .body("Err: database access error".to_owned())
-                            .into_body(),
-                    ),
-                ));
-            }
-        };
-        let uri = &sreq.uri();
-        let col_result = CollectionParam::extrude(&uri, &mut sreq.extensions_mut(), &tags);
-        let collection = match col_result {
-            Ok(v) => v.map(|c| c.collection),
-            Err(e) => {
-                warn!("⚠️ Collection Error:  {:?}", e);
-                queue_report(sreq.extensions_mut(), &e);
-                return Box::pin(future::ok(
-                    sreq.into_response(
+                    ));
+                }
+            };
+
+            let uri = &sreq.uri();
+            let col_result = CollectionParam::extrude(&uri, &mut sreq.extensions_mut(), &tags);
+            let collection = match col_result {
+                Ok(v) => v.map(|c| c.collection),
+                Err(e) => {
+                    warn!("⚠️ Collection Error:  {:?}", e);
+                    queue_report(sreq.extensions_mut(), &e);
+                    return Ok(sreq.into_response(
                         HttpResponse::InternalServerError()
                             .content_type("application/json")
                             .body("Err: bad collection".to_owned())
                             .into_body(),
-                    ),
+                    ));
+                }
+            };
+            let bso = BsoParam::extrude(sreq.head(), &mut sreq.extensions_mut()).ok();
+            let bso_opt = bso.map(|b| b.bso);
+
+            let (req, payload) = sreq.into_parts();
+            let transaction_pool = DbTransactionPool::extract(&req).await?;
+            let resource_ts = transaction_pool
+                .transaction(|db| async move {
+                    db.extract_resource(user_id, collection, bso_opt)
+                        .await
+                        .map_err(Into::into)
+                })
+                .await?;
+            let sreq = ServiceRequest::from_parts(req, payload)
+                .unwrap_or_else(|_| panic!("TODO: this should never happen"));
```

Comment:
Isn't there an `unreachable!()` macro for this?

Reply:
Ya... I'm pretty sure this should never happen, but I also wanted to leave it as a TODO: #695